lxd/device/gpu_sriov.go

```diff
@@ -0,0 +1,319 @@
+	}
+
+	if d.config["pci"] != "" {
+		for _, field := range []string{"id", "productid", "vendorid"} {
+			if d.config[field] != "" {
+				return fmt.Errorf(`Cannot use %q when when "pci" is set`, field)
+			}
+		}
+
+		// PCI devices can be specified as "0000:XX:XX.X" or "XX:XX.X".
+		// However, the devices in /sys/bus/pci/devices use the long format which
+		// is why we need to make sure the prefix is present.
+		if len(d.config["pci"]) == 7 {
+			d.config["pci"] = fmt.Sprintf("0000:%s", d.config["pci"])
+		}
+	}
+
+	if d.config["id"] != "" {
+		for _, field := range []string{"pci", "productid", "vendorid"} {
+			if d.config[field] != "" {
+				return fmt.Errorf(`Cannot use %q when when "id" is set`, field)
+			}
+		}
+	}
+
+	return nil
+}
+
+// validateEnvironment checks the runtime environment for correctness.
+func (d *gpuSRIOV) validateEnvironment() error {
+	return validatePCIDevice(d.config)
+}
+
+// Start is run when the device is added to the instance.
+func (d *gpuSRIOV) Start() (*deviceConfig.RunConfig, error) {
+	err := d.validateEnvironment()
+	if err != nil {
+		return nil, err
+	}
+
+	runConf := deviceConfig.RunConfig{}
+	saveData := make(map[string]string)
+
+	// Get SRIOV parent, i.e. the actual GPU.
+	parentPCIAddress, err := d.getParentPCIAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get PCI information about the GPU device.
+	devicePath := filepath.Join("/sys/bus/pci/devices", parentPCIAddress)
+
+	pciParentDev, err := pci.ParseUeventFile(filepath.Join(devicePath, "uevent"))
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get PCI device info for GPU %q", parentPCIAddress)
+	}
+
+	vfID, err := d.findFreeVirtualFunction(pciParentDev)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to find free virtual function")
+	}
+
+	vfPCIDev, err := d.setupSriovParent(vfID, saveData)
+	if err != nil {
+		return nil, err
+	}
+
+	err = d.volatileSet(saveData)
+	if err != nil {
+		return nil, err
+	}
+
+	runConf.GPUDevice = append(runConf.GPUDevice, []deviceConfig.RunConfigItem{
+		{Key: "devName", Value: d.name},
+		{Key: "pciSlotName", Value: vfPCIDev.SlotName},
+	}...)
+
+	return &runConf, nil
+}
+
+// getParentPCIAddress returns the PCI address of the parent GPU.
+func (d *gpuSRIOV) getParentPCIAddress() (string, error) {
+	gpus, err := resources.GetGPU()
+	if err != nil {
+		return "", err
+	}
+
+	var parentPCIAddress string
+
+	for _, gpu := range gpus.Cards {
+		// Skip any cards that don't match the vendorid, pci, productid or DRM ID settings (if specified).
+		if (d.config["vendorid"] != "" && gpu.VendorID != d.config["vendorid"]) ||
+			(d.config["pci"] != "" && gpu.PCIAddress != d.config["pci"]) ||
+			(d.config["productid"] != "" && gpu.ProductID != d.config["productid"]) ||
+			(d.config["id"] != "" && (gpu.DRM == nil || fmt.Sprintf("%d", gpu.DRM.ID) != d.config["id"])) {
+			continue
+		}
+
+		if parentPCIAddress != "" {
+			return "", fmt.Errorf("VMs cannot match multiple GPUs per device")
```

Comment:
@monstermunchkin I would drop the "VMs" part of this error message, as the device is restricted to VMs only (currently) anyway, and avoids needing to change it in the future if we add support for other instance types.